File: Longest Substring Without Repeating.java
Given a string s, find the length of the longest substring without repeating characters.

Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.

Input: s = "dvdfadbf"
Output: 4
Explanation: The answer is "vdfa", with the length of 4.


class Solution 
{
    public int lengthOfLongestSubstring(String s) 
    {
        int lp = 0;
        int rp = 0;
        int max = 0;
        Set<Character> set = new HashSet<>();            ------------------->  Without repeating means 'Uniqueness'.
        
        while(rp < s.length())
        {
            if( !( set.contains(s.charAt(rp)) ) )        ------------------->  If the set does not contain the current character, add it to the set and move the pointer.
            {
                set.add(s.charAt(rp));
                rp++;
                
                if(set.size() > max) max = set.size();    ----------------> Update the max length if necessary.
            }
            else      -----------------------> If the set does contain the current character, remove every single character comes before and including the repeating character. 
            {
                set.remove(s.charAt(lp));
                lp++;
            }
        }
        return max;
    }
}

//----------------------------------------------------------------------------------------------------------------------
class Solution 
{
    public int lengthOfLongestSubstring(String s) 
    {
          int cur = 0;
          int max = 0;
          StringBuilder sb = new StringBuilder();       ------------> Another solution using StringBuilder.

          for(int i=0;i<s.length();i++)    ----------------> Go through the input string. If the StringBuilder contains the current character, delete every character comes before and including the repeating character.
          {
              if(sb.toString().contains(s.charAt(i) + ""))
              {
                 if(cur > max) max = cur;
                  
                  int occur_index = sb.lastIndexOf(s.charAt(i) + "");
                  sb.delete(0,occur_index + 1);
                  
                  cur = cur - (occur_index + 1);
              }
                  sb.append(s.charAt(i));                      ---------------------> No matter what we'll append the current character to the StringBuilder.
                  cur++;
          }
        
        if(cur > max) max = cur;
        return max;
    }
}

File: Longest Substring Without Repeating.py
https://leetcode.com/problems/longest-substring-without-repeating-characters/submissions/
  
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        unique_chars = set()
        l, res = 0, 0
        
        for r in range(len(s)):
            while s[r] in unique_chars:
                unique_chars.remove(s[l])
                l += 1
                
            unique_chars.add(s[r])
            res = max(res, r - l + 1)
        
        return res
-------------------------------------------------------------------------------------------------------------------------------------------------------------
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if 0 <= len(s) <= 1:
            return len(s)
        
        l, r, max_length = 0, 0, 0
        character_set = set()
        
        # Sliding window question:
        # If the current char previously showed up and was added to the set, we remove all the characters before and including that character from the set.
        # Otherwise, just add the current char to the set and compare the size.
        while r < len(s):
            if s[r] in character_set:
                character_set.remove(s[l])
                l += 1
            else:
                character_set.add(s[r])
                r += 1
                max_length = max(max_length, len(character_set))
                
        return max_length
